044-kraft-liveness-readiness.md

```diff
@@ -0,0 +1,198 @@
+
+To check that the server is listening on a particular port we will use `netstat`, e.g.:
+```shell
+netstat -lnt | grep -Eq 'tcp6?[[:space:]]+[0-9]+[[:space:]]+[0-9]+[[:space:]]+[^ ]+:9091.*LISTEN[[:space:]]*'
+```
+
+To check that the BrokerState metric has the correct value we will use the existing mechanism of checking for the existence of 
+a file on disk that is created by the KafkaAgent.
+In future this could be replaced with a call to an endpoint, since other proposals discuss adding an endpoint that returns this metric.
+
+### Impact on the KafkaRoller
+
+The existing KafkaRoller checks whether a pod is marked as ready to determine whether it needs to be rolled.
+This relies on the fact that currently the pod being ready implies that the BrokerState is >= RUNNING.
+This means any change to the readiness probes also impacts the KafkaRoller.
+In future it would be better for the KafkaRoller to do its own checks, rather than relying on the readiness probes implementation.
+
+The details for how the KafkaRoller should work going forwards should be discussed in a separate proposal.
+This section describes some ideas for how the KafkaRoller should work in KRaft mode based on investigations done while writing this proposal.
+It can be used as a starting point for the KafkaRoller in KRaft mode proposal.
+The key requirement for this proposal to be fully implemented is that the KafkaRoller takes into account the state of controller pods when deciding whether to roll broker pods.
+
+Based on investigations as part of writing this proposal, in KRaft mode:
+* The KafkaRoller should observe the BrokerState metric and any other metrics it needs directly, rather than inferring the state based on readiness.
+* If more than one controller pod has not become ready, KafkaRoller should try to roll the controllers that aren't the current active controller first.
+* If a broker pod has not become ready, KafkaRoller should check the controller quorum is formed before rolling the pod.
+* If a combined pod has not become ready, KafkaRoller should check that all other combined pods have been scheduled 
+(i.e. not in pending state) before rolling the pod or waiting for it to become ready.
+
+The reason for the final bullet (combined pod not being ready) might not be immediately obvious, so it is explained below.
+The brokers do not move to RUNNING until a quorum has been formed.
+For example in combined mode during normal startup the following would happen:
+
+* All the pods are started at the same time
+* The controller processes in each pod form a quorum
+* The broker processes start up and start talking to the controller quorum so the controller unfences them and they move to RUNNING and are marked as ready
+* All pods are now ready
+
+However, if the pods are started initially with affinity constraints so they are all in pending, then the constraint is removed, the following would happen:
```

Comment:
```suggestion
However, if the pods are started with affinity constraints, they are all in a pending state and the following issue arises when forming a quorum:
```

Reply:
I was a bit confused by the _constraint is removed_. I assume we're saying how the final bullet is handled with affinity constraints. Are we saying this is a limitation that needs resolving with KafkaRoller first?  Maybe we should make any dependencies more explicit?

Reply:
Not quite, so the below bullets describe what happens if you initially have constraints that mean the pods can't be scheduled, and then later remove that constraint. The result is basically that even though the cluster should come up, it doesn't due to the way KafkaRoller works. I've made some changes to hopefully make it clearer, see what you think.

Reply:
Thanks @katheris. I understand now.